pulp_gem/tests/functional/api/test_full_mirror.py

```diff
@@ -0,0 +1,77 @@
+import pytest
+import subprocess
+from aiohttp.client_exceptions import ClientResponseError
+
+
+def test_pull_through_metadata(
+    gem_remote_factory,
+    gem_distribution_factory,
+    gem_content_api_client,
+    artifacts_api_client,
+    http_get,
+):
+    """
+    Test that pull-through caching can retrieve metadata files upstream, but does not save them.
+    """
+    artifacts_before = artifacts_api_client.list().count
+    content_before = gem_content_api_client.list().count
+
+    # Choose a remote source that supports all the different gem repository metadata formats
+    remote = gem_remote_factory(url="https://rubygems.org")
+    distribution = gem_distribution_factory(remote=remote.pulp_href)
+
+    urls = [
+        "info/pulp_file_client",
+        "specs.4.8",
+        "latest_specs.4.8",
+        "prerelease_specs.4.8",
+        "quick/Marshal.4.8/pulp_file_client-1.14.0.gemspec.rz",
+    ]
+    for path in urls:
+        url = distribution.base_url + path
+        assert http_get(url)
+
+    artifacts_after = artifacts_api_client.list().count
+    content_after = gem_content_api_client.list().count
+
+    assert artifacts_before == artifacts_after
+    assert content_before == content_after
+
+    with pytest.raises(ClientResponseError) as e:
+        http_get(distribution.base_url + "NOT_A_VALID_LINK")
+
+    assert e.value.status == 404
+
+
+def test_pull_through_install(
+    gem_remote_factory, gem_distribution_factory, gem_content_api_client, delete_orphans_pre
+):
+    """
+    Test that gem clients can install from a distribution with pull-through caching.
+    """
+    out = subprocess.run(("which", "gem"))
+    if out.returncode != 0:
+        pytest.skip("gem not installed on test machine")
+    content_before = gem_content_api_client.list().count
+
+    remote = gem_remote_factory(url="https://rubygems.org")
```

Comment:
Where's the repository to receive all the cached content?

Reply:
You don't need it. Pulp itself is the repository that is storing the cached content. An admin user would need to list the content and manually add it to a repository if they want to.

We could add a feature request in pulpcore to try to add the pull-through content to an attached repository.

Reply:
So the caching is actually only done until someone calls orphan cleanup.
Yes, I think we should handle the case when a pull-through distribution has a repository and a remote. You could use that e.g. to collect all the packages needed to build one container and then move it into an immutable set.

Reply:
Not this PR then...